internal/liveness/liveness.go

```diff
@@ -37,57 +44,113 @@ var (
+	}
+	util.ExtendedLog(ctx, "Healthz req: Health check succeeded")
+}
+
+func getLiveness(c *probeConn) {
+	ctx, cancel := context.WithTimeout(context.Background(), c.config.ProbeTimeout)
+	defer cancel()
+
+	util.TraceLog(ctx, "Metrics req: Sending probe request to CSI driver: %s", c.config.DriverName)
+	ready, err := rpc.Probe(ctx, c.conn)
 	if err != nil {
 		liveness.Set(0)
-		util.ErrorLogMsg("health check failed: %v", err)
+		util.ErrorLog(ctx, "Metrics req: health check failed: %v", err)
 		return
 	}
 
 	if !ready {
 		liveness.Set(0)
-		util.ErrorLogMsg("driver responded but is not ready")
+		util.ErrorLog(ctx, "Metrics req: driver responded but is not ready")
 		return
 	}
 	liveness.Set(1)
-	util.ExtendedLogMsg("Health check succeeded")
+	util.ExtendedLog(ctx, "Metrics req: Health check succeeded")
 }
 
-func recordLiveness(endpoint, drivername string, pollTime, timeout time.Duration) {
-	liveMetricsManager := metrics.NewCSIMetricsManager(drivername)
+func recordLiveness(c *probeConn) {
 	// register prometheus metrics
 	err := prometheus.Register(liveness)
 	if err != nil {
 		util.FatalLogMsg(err.Error())
 	}
 
-	csiConn, err := connlib.Connect(endpoint, liveMetricsManager)
-	if err != nil {
-		// connlib should retry forever so a returned error should mean
-		// the grpc client is misconfigured rather than an error on the network
-		util.FatalLogMsg("failed to establish connection to CSI driver: %v", err)
-	}
-
 	// get liveness periodically
-	ticker := time.NewTicker(pollTime)
+	ticker := time.NewTicker(c.config.PollTime)
 	defer ticker.Stop()
 	for range ticker.C {
-		getLiveness(timeout, csiConn)
+		getLiveness(c)
 	}
 }
 
 // Run starts liveness collection and prometheus endpoint.
 func Run(conf *util.Config) {
 	util.ExtendedLogMsg("Liveness Running")
 
+	liveMetricsManager := metrics.NewCSIMetricsManager("")
+
+	csiConn, err := connlib.Connect(conf.Endpoint, liveMetricsManager)
+	if err != nil {
+		// connlib should retry forever so a returned error should mean
+		// the grpc client is misconfigured rather than an error on the network
+		util.FatalLogMsg("failed to establish connection to CSI driver: %v", err)
+	}
+
+	conf.DriverName, err = rpc.GetDriverName(context.Background(), csiConn)
+	if err != nil {
+		util.FatalLogMsg("failed to get CSI driver name: %v", err)
+	}
+	liveMetricsManager.SetDriverName(conf.DriverName)
+	util.ExtendedLogMsg("CSI driver: %s, Endpoint: %s", conf.DriverName, conf.Endpoint)
+
+	pc := &probeConn{
+		config: conf,
+		conn:   csiConn,
+	}
+
 	// start liveness collection
-	go recordLiveness(conf.Endpoint, conf.DriverName, conf.PollTime, conf.PoolTimeout)
+	go recordLiveness(pc)
 
 	// start up prometheus endpoint
 	util.StartMetricsServer(conf)
+
+	address := net.JoinHostPort(conf.MetricsIP, conf.HealthzPort)
+	http.HandleFunc(conf.HealthzPath, pc.checkProbe)
+	util.ExtendedLogMsg("Serving Health requests on: http://%s%s", address, conf.HealthzPath)
+	err = http.ListenAndServe(address, nil)
```

Comment:
why we are starting one more server. is it needed?

Reply:
Yeah, this is only needed if we want to listen on separate ports.
For a single port and multiple paths like: '/metrics' and '/healthz', we can live with a single server

Reply:
I would suggest to start one server not multiple servers